examples/gagent-proxy-redundant-downstream-integration.py

```diff
@@ -0,0 +1,177 @@
+
+
+def status(juju_statuses):
+    """Status assertion for a cyclic relation between cos-proxy, grafana-agent, and prometheus.
+
+    >>> status({"failing-status": test_invalid_status()})  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+    ...
+    AssertionError: Remove the relation between ... (cos-proxy) and prometheus. ...
+
+    >>> status({"passing-status": test_valid_status()})
+    """
+    gagent_and_proxy_rel = False
+    suspicious_endpoint_apps = {}
+    for status_name, status in juju_statuses.items():
+        applications = status.get("applications", {})
+
+        # Gather suspicious grafana-agent relations to prometheus
+        if not (agent := get_app_by_charm_name(status, "grafana-agent")):
+            continue
+        for endpoint, relations in agent.get("relations", {}).items():
+            for rel in relations:
+                if endpoint == "cos-agent":
+                    if applications.get(rel["related-application"])["charm"] == "cos-proxy":
+                        gagent_and_proxy_rel = True
+                elif endpoint == "send-remote-write":
+                    suspicious_endpoint_apps.setdefault(endpoint, [])
+                    suspicious_endpoint_apps[endpoint].append(rel["related-application"])
+
+        # Gather suspicious cos-proxy relations to prometheus
+        if not (proxy := get_app_by_charm_name(status, "cos-proxy")):
+            continue
+        for endpoint, relations in proxy.get("relations", {}).items():
+            if endpoint != "downstream-prometheus-scrape":
+                continue
+            for rel in relations:
+                suspicious_endpoint_apps.setdefault(endpoint, [])
+                suspicious_endpoint_apps[endpoint].append(rel["related-application"])
+
+        # Assert that the suspicious relations are not redundant
+        for app in suspicious_endpoint_apps.get("downstream-prometheus-scrape", {}):
+            redundant_downstream = app in suspicious_endpoint_apps.get("send-remote-write", {})
+            ga_app = get_app_name_by_charm_name(status, "grafana-agent")
+            cp_app = get_app_name_by_charm_name(status, "cos-proxy")
+            assert not (gagent_and_proxy_rel and redundant_downstream), (
+                f'Remove the relation between "{cp_app}" (cos-proxy) and prometheus. "{cp_app}" '
+                f'(cos-proxy) and "{ga_app}" (grafana-agent) are inter-related (cos-agent) and '
+                f'related to the same prometheus in "{status_name}"'
+            )
+
+
+# ==========================
+# Helper methods
+# ==========================
+
+
+def get_app_name_by_charm_name(status: dict, charm_name: str) -> Optional[str]:
+    """Helper function to get the (unpredictable) application name from a charm name."""
+    if applications := status.get("applications", {}):
+        for app, context in applications.items():
+            if charm_name == context["charm"]:
+                return app
+    return None
```

Comment:
The return type should be a list, not a scalar, because we can have different app names with the same charm type.

Is it safe enough to assume that "applications" key will always be there?

```python
    return [app_name for app_name, context in status["applications"].items() if context["charm"] == charm_name] or None
```

But since an empty list evaluates as False, do we need the Optional?